feat: Responsive sidebar

henit-chobisa opened the pull request:
# Summary

## Additions : 
- Added `UIContext` for Universal State Management of the Sidebar.
- Added Conditional Rendering to the Sidebar.
- Added new `Navbar Item`, `menu` button, that toggles the sidebar on smaller window sizes.

## Effect : 
- UI Context Added for further UI State Management.
- Responsive Sidebar would be implemented in the Config-UI.
- There would be a toggle button for seen in the smaller screen sizes.


### Does this close any open issues?
Closes #3106 

### Screenshots
![devlakeSidebarSolution](https://user-images.githubusercontent.com/72302948/190841597-bc79e707-69a5-4e6c-b880-cca9694bbe96.gif)
![devlakeSidebarSolution2](https://user-images.githubusercontent.com/72302948/190841633-c38bb882-bf8f-4481-b952-c49a3614cf1e.gif)


### Other Information
Any other information that is important to this PR.

e2corporation commented on config-ui/src/components/Nav.jsx:
```diff
@@ -26,11 +26,41 @@ import {
+  const [menuClass, setMenuClass] = useState("navbarMenuButton");
+  const size = useWindowSize();
+
+  const toggleSidebarOpen = (open) => {
```
I think this could be refactored to be more declarative and drop the need for using the `IF` and `return` statements.
```suggestion
  const toggleSidebarOpen = (open) => {
  uiContext.changeSidebarVisibility(open)
  setMenuClass(open ? 'navbarMenuButtonSidebarOpened' : 'navbarMenuButton'
 }
```

e2corporation commented on config-ui/src/components/Nav.jsx:
```diff
@@ -26,11 +26,41 @@ import {
+    setMenuClass("navbarMenuButton");
+  }
+
+  useEffect(() => {
```
This Effect could also be refactored to be more declarative and remove the nested `IF` statements
```suggestion
  useEffect(() => {
        toggleSidebarOpen(size.width >= 850 && uiContext.sidebarVisible != true ? true : false)
  }, [size]);
```

e2corporation commented on config-ui/src/components/Nav.jsx:
```diff
@@ -26,11 +26,41 @@ import {
+  }
+
+  useEffect(() => {
+    if (size.width >= 850 ) {
```
Would it be beneficial to store the `850` breakpoint value as `desktopBreakpointWidth` within the UIContext's state object so it can be configurable?

e2corporation commented on config-ui/src/hooks/useWIndowSize.jsx:
```diff
@@ -0,0 +1,29 @@
+import { useEffect, useState } from "react";
+
+function useWindowSize() {
+    const [windowSize, setWindowSize] = useState({
```
Would it be better to initialize these to integer values rather than undefined, or is undefined desired? Maybe even a min-viewport size of 320x240.
```suggestion
    const [windowSize, setWindowSize] = useState({
      width: 320,
      height: 240,
    })
```

e2corporation commented on config-ui/src/styles/nav.scss:
```diff
@@ -17,11 +17,43 @@
   z-index: 1;
 }
 
+.navbarMenuButton{
```
This css rule is missing a space before the opening bracket `{`
```suggestion
.navbarMenuButton {
```

e2corporation commented on config-ui/src/styles/nav.scss:
```diff
@@ -17,11 +17,43 @@
+  }
+}
+
+.navbarMenuButtonSidebarOpened{
```
This css rule is missing a space before the opening bracket `{`
```suggestion
.navbarMenuButtonSidebarOpened {
```

henit-chobisa commented:
Thank You @e2corporation for such a detailed Review,
I will cover up on all your points and push the fixtures by tomorrow EOD

e2corporation commented:
> Thank You @e2corporation for such a detailed Review, I will cover up on all your points and push the fixtures by tomorrow EOD

Great work so far Henit! I'd like to run some functional tests later as well since I had some questions with the resize handling. I see the cleanup function has been added to the resize effect which is nice and will prevent memory leaks. One caveat of wrapping the whole app in a Context too is that it may cause unwanted re-renders which could impact form control components, so I just wanted to make sure there aren't any potential side effects.

I also had another idea to extend this feature, which is to create a "Docked" mode for the sidebar (slim bar with DevLake logo + icons), which maybe something you might be interested in implementing.

henit-chobisa commented:
> 
@e2corporation 

Sure, Julien!!
I was also a bit scared of wrapping the whole app with the `UIContext` 😅 but I couldn't find any alternative way
If I would have wrapped it with the page, still the amount of renders wouldn't have reduces.

What an amazing idea!!, Julien, this is why I love open source, Creativity. 😍
Docks are trendy these day, we can also give our application that modern look, and I also look forward implementing something in the dock which gives it that unique look which people might copy! ⚡️
I would love to work on this 🫡

e2corporation commented on config-ui/src/components/Nav.jsx:
```diff
@@ -26,11 +26,30 @@ import {
 import { ReactComponent as SlackIcon } from '@/images/slack-mark-monochrome-black.svg'
 import { ReactComponent as SlackLogo } from '@/images/slack-rgb.svg'
+import UIContext from '../store/UIContext'
+import useWindowSize from '../hooks/useWIndowSize'
```
```suggestion
import useWindowSize from '@/hooks/useWIndowSize'
```

e2corporation commented on config-ui/src/index.js:
```diff
@@ -20,8 +20,11 @@ import React from 'react'
 import ReactDOM from 'react-dom'
 
 import App from './App'
+import { UIContextProvider } from './store/UIContext'
```
```suggestion
import { UIContextProvider } from '@/store/UIContext'
```

e2corporation commented:
> Thank You @e2corporation for such a detailed Review, I will cover up on all your points and push the fixtures by tomorrow EOD

Thanks for applying the updates Henit, we also did much refactoring on `main` branch recently to apply code formatting standards to the UI (`@config-ui`). This PR will need to be rebased, or to prevent conflicts you can re-create a new branch from `main` and cherry-pick all these changes to that new branch and re-force push.

e2corporation commented:
> @e2corporation Julien, would it be fine if I go with resolving the conflict, are there any downsides to that ? <img alt="image" width="923" src="https://user-images.githubusercontent.com/72302948/191818737-6916f460-907e-4ab6-b322-668697b627af.png">

I'm not sure if/how Github will auto-resolve the conflicts successful given the nature of the recent changes, so resolving the conflicts may still end up being a manual process.

henit-chobisa commented:
Ok sure, understood
I will rebase the branch and cherry pick the changes!
Thank You for keeping up with me and guiding me at every step.

henit-chobisa commented:
@e2corporation 
Julien, I have worked on what you said and now I think that the branch is up to date with master.

e2corporation commented:
@henit-chobisa Thanks for the updates Henit, it does look like we have some unwanted **backend** files on this PR that will need to be removed. Can these resources be reset to `HEAD` version from `main` so they are discarded from this PR?

- `Makefile`
- `plugins/`

henit-chobisa commented:
> @henit-chobisa Thanks for the updates Henit, it does look like we have some unwanted **backend** files on this PR that will need to be removed. Can these resources be reset to `HEAD` version from `main` so they are discarded from this PR?
> 
> * `Makefile`
> * `plugins/`

Julien, I have made the changes you asked for 🫡

e2corporation commented:
@henit-chobisa Thanks Henit, can you also resolve the ESLint errors mentioned on the **Frontend-Lint** Github CI Check? There are some missing effect dependencies. https://github.com/apache/incubator-devlake/actions/runs/3129884127/jobs/5079735842

In regards to the error of `bind` usage, Instead of using bind, wrap the handler in a closure so there won't be a need for using bind `() => {}`

<img width="1107" alt="Screen Shot 2022-09-26 at 2 09 33 PM" src="https://user-images.githubusercontent.com/1742233/192349518-0860096c-aeb9-42be-a93b-13374838945d.png">

e2corporation commented on config-ui/src/components/Nav.jsx:
```diff
@@ -15,73 +15,104 @@
+      <Navbar.Group className={menuClass}>
+        <Icon
+          icon={uiContext.sidebarVisible ? 'menu-closed' : 'menu-open'}
+          onClick={toggleSidebarOpen.bind(null, !uiContext.sidebarVisible)}
```
In order to resolve one of the ESLint Errors, this can be refactored to:
```suggestion
          onClick={(e) => toggleSidebarOpen(!uiContext.sidebarVisible)}
```

e2corporation commented:
@henit-chobisa Additionally, the commit message structure for the last 2 commits does not conform to the current linter format (`fix:` `feat:` `refactor:` `chore:` etc) and is also causing the **Commit Message** CI Check to fail, can those commits messages be amended as well to fit the desired format?

<img width="577" alt="Screen Shot 2022-09-26 at 2 16 26 PM" src="https://user-images.githubusercontent.com/1742233/192351052-09c46ece-ec84-42f9-a70a-b48656493980.png">

**Expected format to pass CI Check**
- chore: removed makefile
- chore: removed blueprint.go

**or**
- refactor: removed makefile
- refactor: removed blueprint.go

henit-chobisa commented:
> @henit-chobisa Additionally, the commit message structure for the last 2 commits does not conform to the current linter format (`fix:` `feat:` `refactor:` `chore:` etc) and is also causing the **Commit Message** CI Check to fail, can those commits messages be amended as well to fit the desired format?
> 
> <img alt="Screen Shot 2022-09-26 at 2 16 26 PM" width="577" src="https://user-images.githubusercontent.com/1742233/192351052-09c46ece-ec84-42f9-a70a-b48656493980.png">
> 
> **Expected format to pass CI Check**
> 
> * chore: removed makefile
> * chore: removed blueprint.go
> 
> **or**
> 
> * refactor: removed makefile
> * refactor: removed blueprint.go

I'm extremely sorry, I just forgot that
I fix that now.

e2corporation commented:
@henit-chobisa Another issue that may affect this PR from a clean merge is that you have issued `merge` commits when updating changes from `main` branch, when we prefer a `rebase` policy instead (ie `git pull --rebase`) that way feature commits are replayed on top of main cleanly.

henit-chobisa commented:
> > > @henit-chobisa Additionally, the commit message structure for the last 2 commits does not conform to the current linter format (`fix:` `feat:` `refactor:` `chore:` etc) and is also causing the **Commit Message** CI Check to fail, can those commits messages be amended as well to fit the desired format?
> > > <img alt="Screen Shot 2022-09-26 at 2 16 26 PM" width="577" src="https://user-images.githubusercontent.com/1742233/192351052-09c46ece-ec84-42f9-a70a-b48656493980.png">
> > > **Expected format to pass CI Check**
> > > 
> > > * chore: removed makefile
> > > * chore: removed blueprint.go
> > > 
> > > **or**
> > > 
> > > * refactor: removed makefile
> > > * refactor: removed blueprint.go
> > 
> > 
> > I'm extremely sorry, I just forgot that I fix that now.
> 
> @henit-chobisa No problem, it takes a bit of a adjustment getting used to the Linter on your first PR. Subsequent PRs should be much smoother once you've gone through the process. General workflow is once you are code complete and ready to remove Draft state, `npm run lint` should be performed on `@config-ui` to resolve Lint errors and ensure the formatting standards are applied.

Sure Julien, I will take care that I won't create this much issues for reviewers and do everything perfectly.

henit-chobisa commented:
> @henit-chobisa Another issue that may affect this PR from a clean merge is that you have issued `merge` commits when updating changes from `main` branch, when we prefer a `rebase` policy instead (ie `git pull --rebase`) that way feature commits are replayed on top of main cleanly.

So should I undo the merge commits and do rebase with that ?

e2corporation commented:
> > > > @henit-chobisa Additionally, the commit message structure for the last 2 commits does not conform to the current linter format (`fix:` `feat:` `refactor:` `chore:` etc) and is also causing the **Commit Message** CI Check to fail, can those commits messages be amended as well to fit the desired format?
> > > > <img alt="Screen Shot 2022-09-26 at 2 16 26 PM" width="577" src="https://user-images.githubusercontent.com/1742233/192351052-09c46ece-ec84-42f9-a70a-b48656493980.png">
> > > > **Expected format to pass CI Check**
> > > > 
> > > > * chore: removed makefile
> > > > * chore: removed blueprint.go
> > > > 
> > > > **or**
> > > > 
> > > > * refactor: removed makefile
> > > > * refactor: removed blueprint.go
> > > 
> > > 
> > > I'm extremely sorry, I just forgot that I fix that now.
> > 
> > 
> > @henit-chobisa No problem, it takes a bit of a adjustment getting used to the Linter on your first PR. Subsequent PRs should be much smoother once you've gone through the process. General workflow is once you are code complete and ready to remove Draft state, `npm run lint` should be performed on `@config-ui` to resolve Lint errors and ensure the formatting standards are applied.
> 
> Sure Julien, I will take care that I won't create this much issues for reviewers and do everything perfectly.

Thanks Henit, it's ok if issues occur, this is why we have the code review process.

e2corporation commented:
> > @henit-chobisa Another issue that may affect this PR from a clean merge is that you have issued `merge` commits when updating changes from `main` branch, when we prefer a `rebase` policy instead (ie `git pull --rebase`) that way feature commits are replayed on top of main cleanly.
> 
> So should I undo the merge commits and do rebase with that ?

It still looks like it would be a manual cherry-pick process to be able to resolve the issue. I would create a local clone/backup of this branch. Create a new branch from `main` as the starting point, cherry-pick all relevant changes to working copy and then force-push the branch to `henit-chobisa:Responsive-Sidebar`. This PR can also be closed and we can re-issue a new PR from another branch as well.

henit-chobisa commented:
Okay sure @e2corporation, I do the same.
